### PR TITLE
Fix routing of uninitialized values during multipple simultaneous node connect

### PR DESCRIPTION
**Pull request by sander-visser**
… parsing own definition for init value if needed

Also make new provide port data be read as per definition until out file got opened

### CODE REVIEW

File: apx/common/src/apx_nodeInfo.c [hunk 1]
```diff
@@ -492,6 +492,10 @@ void apx_nodeInfo_copyInitDataFromProvideConnectors(apx_nodeInfo_t *self)
                   }
                   else
                   {
+                     if (requireNodeData->inPortDirtyFlags != 0)
+                     {
+                        requireNodeData->inPortDirtyFlags[requirePortEntry->offset] = 1;
```
Comment (sander-visser) [inline apx/common/src/apx_nodeInfo.c:499]: Maybe setting the flag should be done after the data is copied...

File: apx/common/src/apx_nodeManager.c [hunk 2]
```diff
@@ -537,6 +538,8 @@ static void apx_nodeManager_createNode(apx_nodeManager_t *self, const uint8_t *d
                         assert(nodeData->outPortDirtyFlags);
                         nodeData->outPortDataLen = outPortDataLen;
                         APX_LOG_INFO("[APX_NODE_MANAGER]%s Server opening client file %s[%d,%d]", debugInfoStr, fileName, outDataFile->fileInfo.address, outDataFile->fileInfo.length);
+                        // Prior to a response to the file open request let the data be read as defined by definition
+                        apx_nodeManager_createOutInitData(apxNode, nodeData);
```
Comment (sander-visser) [inline apx/common/src/apx_nodeManager.c]: should probably check returncode and set APX_LOG_ERROR

File: apx/common/src/apx_nodeManager.c [hunk 3]
```diff
@@ -577,10 +576,16 @@ static void apx_nodeManager_createNode(apx_nodeManager_t *self, const uint8_t *d
             {
                apx_router_attachNodeInfo(self->router, nodeInfo);
             }
-            //for all connected require ports copy data from the provide port into our newly create inDataFile buffer
-            apx_nodeInfo_copyInitDataFromProvideConnectors(nodeInfo);
             if (inDataFile != 0)
             {
+               // For all connected require ports copy data from the provide port into our newly create inDataFile buffer
+               apx_nodeInfo_copyInitDataFromProvideConnectors(nodeInfo);
+               // For any reqUire port that did not get flag set - get the initdata from the definition instead
```
Comment (sander-visser) [inline apx/common/src/apx_nodeManager.c]: wierd uppercase "U"

File: apx/common/src/apx_nodeManager.c [hunk 4]
```diff
@@ -697,30 +702,68 @@ static void apx_nodeManager_removeNodeInfo(apx_nodeManager_t *self, apx_nodeInfo
...
       portData = adt_bytearray_new(0);
       numRequirePorts = apx_node_getNumRequirePorts(node);
       for(i=0; i<numRequirePorts; i++)
       {
          int32_t packLen;
-         int32_t dataLen;
          apx_port_t *port = apx_node_getRequirePort(node, i);
          assert(port != 0);
          packLen = apx_port_getPackLen(port);
+         if (nodeData->inPortDirtyFlags[offset] == 0)
```
Comment (sander-visser) [inline apx/common/src/apx_nodeManager.c:725]: Add a comment // If no provider is connected then get the init value from own description file

File: apx/common/src/apx_nodeManager.c [hunk 5]
```diff
@@ -697,30 +702,68 @@ static void apx_nodeManager_removeNodeInfo(apx_nodeManager_t *self, apx_nodeInfo
...
+      numProvidePorts = apx_node_getNumProvidePorts(node);
+      for(i=0; i<numProvidePorts; i++)
+      {
+         int32_t packLen;
+         int32_t dataLen;
+         apx_port_t *port = apx_node_getProvidePort(node, i);
+         assert(port != 0);
+         packLen = apx_port_getPackLen(port);
          apx_node_fillPortInitData(node, port, portData);
          dataLen = adt_bytearray_length(portData);
```
Comment (sander-visser) [inline apx/common/src/apx_nodeManager.c]: remove dataLen and move this row into the assert

File: apx/common/src/apx_nodeManager.c [hunk 6]
```diff
@@ -560,11 +563,7 @@ static void apx_nodeManager_createNode(apx_nodeManager_t *self, const uint8_t *d
                assert(nodeData->inPortDataBuf);
                nodeData->inPortDirtyFlags = (uint8_t*) malloc(inPortDataLen);
                assert(nodeData->inPortDirtyFlags);
-               result = apx_nodeManager_createInitData(apxNode, nodeData->inPortDataBuf, inPortDataLen);
-               if (result == false)
-               {
-                  APX_LOG_ERROR("[APX_NODE_MANAGER] Failed to create init data for node %s", apx_node_getName(apxNode));
-               }
+               memset(nodeData->inPortDirtyFlags, 0, inPortDataLen);
```
Comment (jenspommer) [inline apx/common/src/apx_nodeManager.c:568]: Should outPortDirtyFlags also be cleared in the same manner somewhere?

**Comment by sander-visser** [comment]
Fixed my own review comments....

**Comment by sander-visser** [comment]
Fixed my own review comments and tested apx-server function with 5 connected nodes.

In my usecase the apx-server previously used 647 ms CPU time after initial node connections (@60s uptime), and now uses 644 ms.

The in port optimizations seem to offset the initializing of out port from definition during connect

**Pull request merged**